EventFilter/HcalRawToDigi/plugins/HcalDigiToRawuHTR.cc

```diff
@@ -0,0 +1,285 @@
+    const HBHEDigiCollection& qie8hbhedc=*(hbheDigiCollection);
+    for(HBHEDigiCollection::const_iterator qiedf=qie8hbhedc.begin();qiedf!=qie8hbhedc.end();qiedf++){
+      DetId detid = qiedf->id();
+
+      HcalElectronicsId eid(readoutMap->lookup(detid));
+      int crateId = eid.crateId();
+      int slotId = eid.slot();
+      int uhtrIndex = (crateId&0xFF) | ((slotId&0xF)<<8) ; 
+
+      if( ! uhtrs.exist(uhtrIndex) ){
+	uhtrs.newUHTR( uhtrIndex );
+      }
+      uhtrs.addChannel(uhtrIndex,qiedf,readoutMap,_verbosity);
+    }
+  }
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // TP data
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - 
+  // loop over each digi and allocate memory for each
+  if(tpDigiCollection.isValid()){
+    const HcalTrigPrimDigiCollection& qietpdc=*(tpDigiCollection);
+    for(HcalTrigPrimDigiCollection::const_iterator qiedf=qietpdc.begin();qiedf!=qietpdc.end();qiedf++){
+      DetId detid = qiedf->id();
+      HcalElectronicsId eid(readoutMap->lookupTrigger(detid));
+    
+      int crateId = eid.crateId();
+      int slotId = eid.slot();
+      int uhtrIndex = (crateId&0xFF) | ((slotId&0xF)<<8);
+      int ilink = eid.fiberIndex();
+      int itower = eid.fiberChanId();
+      int channelid = (itower&0xF) | ((ilink&0xF)<<4);
+
+      if( ! uhtrs.exist(uhtrIndex) ){
+	uhtrs.newUHTR( uhtrIndex );
+      }
+      uhtrs.addChannel(uhtrIndex,qiedf,channelid,_verbosity);
+    }
+  }
+  // -----------------------------------------------------
+  // -----------------------------------------------------
+  // loop over each uHTR and format data
+  // -----------------------------------------------------
+  // -----------------------------------------------------
+  // loop over each uHTR and format data
+  int idxuhtr =-1;
+  for( UHTRpacker::UHTRMap::iterator uhtr = uhtrs.uhtrs.begin() ; uhtr != uhtrs.uhtrs.end() ; ++uhtr){
+
+    idxuhtr ++;
+   
+    uint64_t crateId = (uhtr->first)&0xFF;
+    uint64_t slotId =  (uhtr->first&0xF00)>>8;
+
+    uhtrs.finalizeHeadTail(&(uhtr->second),_verbosity);
+    int fedId = FEDNumbering::MINHCALuTCAFEDID + crateId;
+    if( fedMap.find(fedId) == fedMap.end() ){
+      /* QUESTION: where should the orbit number come from? */
+      fedMap[fedId] = std::unique_ptr<HCalFED>(new HCalFED(fedId,iEvent.id().event(),iEvent.orbitNumber(),iEvent.bunchCrossing()));
+    }
+    fedMap[fedId]->addUHTR(uhtr->second,crateId,slotId);
+  }// end loop over uhtr containers
+
+  /* ------------------------------------------------------
+     ------------------------------------------------------
+           putting together the FEDRawDataCollection
+     ------------------------------------------------------
+     ------------------------------------------------------ */
+  for( map<int,unique_ptr<HCalFED> >::iterator fed = fedMap.begin() ; fed != fedMap.end() ; ++fed ){
+
+    int fedId = fed->first;
+
+    fed->second->formatFEDdata(fed_buffers->FEDData(fedId));
```

Comment:
about 10 repeated calls to ``` fed_buffers->FEDData(fedId) ```  can be replaced by a temporary
``` auto & fedData =  fed_buffers->FEDData(fedId) ```
Given that addresses are being passed around, I'm not sure that compiler will optimize out this call.